R/cdisc_data.R

```diff
@@ -68,58 +68,71 @@ cdisc_data <- function(...,
-  )
+
+  # todo: is it really important? - to remove
   if (inherits(join_keys, "JoinKeySet")) {
     join_keys <- teal.data::join_keys(join_keys)
   }
 
-  update_join_keys_to_primary(data_objects, join_keys)
-
-  retrieve_parents <- function(x) {
-    tryCatch(
-      x$get_parent(),
-      error = function(cond) rep(character(0), length(x$get_datanames()))
+  if (
+    checkmate::test_list(data_objects, types = c("TealDataConnector", "TealDataset", "TealDatasetConnector"))
+  ) {
+    lifecycle::deprecate_warn(
+      when = "0.3.1",
+      "cdisc_data(
+        data_objects = 'should use data directly. Using TealDatasetConnector and TealDataset is deprecated.'
+      )"
     )
-  }
+    update_join_keys_to_primary(data_objects, join_keys)
+
+    retrieve_parents <- function(x) {
+      tryCatch(
+        x$get_parent(),
+        error = function(cond) rep(character(0), length(x$get_datanames()))
+      )
+    }
+
+    new_parents_fun <- function(data_objects) {
+      lapply(data_objects, function(x) {
+        if (inherits(x, "TealDataConnector")) {
+          unlist(new_parents_fun(x$get_items()), recursive = FALSE)
+        } else {
+          list(retrieve_parents(x))
+        }
+      })
+    }
 
-  new_parents_fun <- function(data_objects) {
-    lapply(data_objects, function(x) {
+    new_parents <- unlist(new_parents_fun(data_objects), recursive = FALSE)
+
+    names(new_parents) <- unlist(lapply(data_objects, function(x) {
       if (inherits(x, "TealDataConnector")) {
-        unlist(new_parents_fun(x$get_items()), recursive = FALSE)
+        lapply(x$get_items(), function(z) z$get_dataname())
       } else {
-        list(retrieve_parents(x))
+        x$get_datanames()
       }
-    })
-  }
+    }))
 
-  new_parents <- unlist(new_parents_fun(data_objects), recursive = FALSE)
-
-  names(new_parents) <- unlist(lapply(data_objects, function(x) {
-    if (inherits(x, "TealDataConnector")) {
-      lapply(x$get_items(), function(z) z$get_dataname())
-    } else {
-      x$get_datanames()
+    if (is_dag(new_parents)) {
+      stop("Cycle detected in a parent and child dataset graph.")
     }
-  }))
+    join_keys$set_parents(new_parents)
+    join_keys$update_keys_given_parents()
 
-  if (is_dag(new_parents)) {
-    stop("Cycle detected in a parent and child dataset graph.")
-  }
-  join_keys$set_parents(new_parents)
-  join_keys$update_keys_given_parents()
+    x <- TealData$new(..., check = check, join_keys = join_keys)
 
-  x <- TealData$new(..., check = check, join_keys = join_keys)
+    if (length(code) > 0 && !identical(code, "")) {
+      x$set_pull_code(code = code)
+    }
 
-  if (length(code) > 0 && !identical(code, "")) {
-    x$set_pull_code(code = code)
+    x$check_reproducibility()
+    x$check_metadata()
+    x
+  } else {
+    if (!checkmate::test_names(names(data_objects), type = "named")) {
+      stop("Dot (`...`) arguments on `teal_data()` must be named.")
+    }
```

Comment:
Do we want to support ["dynamic dots"](https://rlang.r-lib.org/reference/dyn-dots.html)?

I am thinking here about re-collecting dots with `rlang:: dots_list` which

Reply:
Actually I think that this could be a very neat feature implemented at relatively low cost.

Reply:
Yes, but at this moment this list could be named (new approach) and unnamed (old approach). So I can't just assert directly when obtaining `...`

```
  data_objects <- list(...)
  if (checkmate::test_list(data_objects, types = c("TealDataConnector", "TealDataset", "TealDatasetConnector")) {
     # make TealData
   } else {
     if (!checkmate::test_names(names(data_objects), type = "named")) {
      stop("Dot (`...`) arguments on `teal_data()` must be named.")
    }
    # make teal_data
   
   }

```

Reply:
You would have to collect dots twice
```
if (old approach condition) {
  objs <- dots_list(..., .named = FALSE, <other>)
  ...
} else {
  objs <- dots_list(..., .named = TRUE, <other>)
  ...
}
```

Reply:
Three times actually, because logical condition is based on output from `...`

Reply:
ok how about this:
```
objs <- dots_list(..., .named = FALSE, <other>)
if (condition(objs)) {
  ...
} else {
  objs <- dots_list(..., .named = TRUE, <other>)
  ...
}
```

Then after we hard deprecate old approach it would be just this:
```
  objs <- dots_list(..., .named = TRUE, <other>)
  ...
```